k3s-io/k3s
# Improve etcd load-balancer startup behavior

#### Proposed Changes ####

Prefer the address of the etcd member being joined, and seed the full address list immediately on startup.

Previously, the first server listed in `etcdctl member list` would be used as the default server for the load-balancer.  That would cause problems if that node was in the process of being removed from the cluster, and we're intentionally being restarted with a different server address.

With this change the etcd load-balancer should always use the node targeted by the --server address, if that node has the etcd role. If it does not, then the first etcd cluster member will be used.

#### Types of Changes ####

bugfix

#### Verification ####

1. Start 3 etcd-only servers
2. Add a control-plane-only server to the cluster
3. Note that the address of the server specified in the --server address is always reported as the server address in the etcd load-balancer log message:
    ```
    Mar 27 15:17:21 controlnode k3s[540944]: time="2024-03-27T15:17:21Z" level=info msg="Adding server to load balancer k3s-etcd-server-load-balancer: 192.168.86.72:2379"
    Mar 27 15:17:21 controlnode k3s[540944]: time="2024-03-27T15:17:21Z" level=info msg="Running load balancer k3s-etcd-server-load-balancer 127.0.0.1:2379 -> [192.168.86.72:2379] [default: 192.168.86.72:2379]" 
    ```


#### Testing ####

#### Linked Issues ####

* https://github.com/k3s-io/k3s/issues/9882

#### User-Facing Change ####

```release-note

```

#### Further Comments ####
